File: c++/Counting-Sort.cpp
#include <iostream>
#include <vector>
using namespace std;

void countingSort(int arr[], int n) {
    // Find the maximum value in the array
    int maxElement = arr[0];
    for (int i = 1; i < n; i++) {
        if (arr[i] > maxElement) {
            maxElement = arr[i];
        }
    }

    // Create count array of size (maxElement+1)
    vector<int> count(maxElement + 1, 0);

    // Store the count of each element
    for (int i = 0; i < n; i++) {
        count[arr[i]]++;
    }

    // Build the sorted array
    int index = 0;
    for (int i = 0; i <= maxElement; i++) {
        while (count[i] > 0) {
            arr[index] = i;
            index++;
            count[i]--;
        }
    }
}

int main() {
    int n;
    cout << "Enter number of elements: ";
    cin >> n;

    int arr[n];
    cout << "Enter elements (non-negative integers): ";
    for (int i = 0; i < n; i++) cin >> arr[i];

    countingSort(arr, n);

    cout << "Sorted array using Counting Sort: ";
    for (int i = 0; i < n; i++) cout << arr[i] << " ";
    cout << endl;

    return 0;
}

File: go/counting_sort.go
package main

import (
	"fmt"
)

// CountingSort sorts an array of non-negative integers using counting sort algorithm.
func CountingSort(arr []int) []int {
	if len(arr) == 0 {
		return arr
	}

	// Step 1: Find the maximum element in the array
	max := arr[0]
	for _, num := range arr {
		if num > max {
			max = num
		}
	}

	// Step 2: Create a count array
	count := make([]int, max+1)

	// Step 3: Count the occurrences of each element
	for _, num := range arr {
		count[num]++
	}

	// Step 4: Reconstruct the sorted array
	index := 0
	for i, c := range count {
		for c > 0 {
			arr[index] = i
			index++
			c--
		}
	}

	return arr
}

func main() {
	var n int
	fmt.Print("Enter number of elements: ")
	fmt.Scan(&n)

	arr := make([]int, n)
	fmt.Println("Enter elements:")

	for i := 0; i < n; i++ {
		fmt.Scan(&arr[i])
	}

	fmt.Println("Original array:", arr)
	sorted := CountingSort(arr)
	fmt.Println("Sorted array:  ", sorted)
}
